design_docs/2023-02_hydro_cli.md

```diff
@@ -0,0 +1,105 @@
+# Hydro CLI Design
+Hydroflow makes it possible to describe single node/thread stream processors, but is unopinionated when it comes to networking and deployment. Working towards the goal of Hydro being and end-to-end stack for distributed programming, the first step is to automate **deploying** standalone Hydroflow programs to cloud machines and **wiring** them together to form a distributed application. This document describes the design of the Hydro CLI, which will be used to achieve these goals.
+
+The Hydro CLI (`hydro`) allows developers to define Python and JavaScript programs that specify how a set of Hydroflow programs should be deployed. Unlike Terraform, where config files generate a static JSON spec, and SkyPilot, which uses YAML config to specify constraints on machines, Hydro config files are **dynamic**. A single config file can be used to generate multiple deployments, each with different parameters. This allows developers to easily experiment with different configurations and conditions. And on top of that, the same config files can interact and monitor with **live** deployments, making it possible to run complex experiments.
+
+We envision the following use cases:
+- Deploying services (like chat apps) to the cloud and exposing them to the internet
+- Running research experiments (like compartmentalized paxos) that involve several configurations and dynamic conditions (like network partitions)
+- Running applications that involve a mix of Hydroflow and non-Hydroflow components (like a web server and a database)
+
+## Simple Example
+Consider the classic echo program example in Hydroflow:
+```rust
+// src/echo_server.hf
+use data_types::EchoMsg; // defined in data_types crate in this workspace
+use chrono::prelude::*;
+use std::net::SocketAddr;
+
+module main(
+    input inbound: (EchoMsg, SocketAddr),
+    output outbound: (EchoMsg, SocketAddr)
+) {
```

Comment:
!!?

Reply:
hahaha, this is just me playing around with what modules in HF could look like... not committed to it yet at all